internal/server/email/smtp.go

```diff
@@ -0,0 +1,238 @@
+		return err
+	}
+	if err = c.writeln(base64encode(SendgridAPIKey)); err != nil {
+		return err
+	}
+	if _, err = c.readln("235 Authentication successful"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func base64encode(s string) string {
+	return base64.StdEncoding.EncodeToString([]byte(s))
+}
+
+// readln reads a line of text ending in \n. if expected is empty it returns the result.
+// if expected is not empty, the read string must match or start with the expected string
+func (c *Client) readln(expected string) (result string, err error) {
+	defer func() {
+		if err != nil {
+			c.Conn = nil
+		}
+	}()
+
+	if err := c.SetDeadline(time.Now().Add(30 * time.Second)); err != nil {
+		return "", fmt.Errorf("set deadline: %w", err)
+	}
+
+	s, err := c.reader.ReadString('\n')
+	if err != nil {
+		return "", err
+	}
+
+	s = strings.TrimRight(s, "\r\n")
+
+	if expected != "" && !strings.HasPrefix(s, expected) {
+		return s, fmt.Errorf("Unexpected value read: %q while waiting for %q", s, expected)
+	}
+
+	return s, nil
+}
+
+func (c *Client) writeln(s string) (err error) {
+	defer func() {
+		if err != nil {
+			c.Conn = nil
+		}
+	}()
+
+	s += "\r\n"
+	if err = c.SetWriteDeadline(time.Now().Add(30 * time.Second)); err != nil {
+		return fmt.Errorf("set write deadline: %w", err)
+	}
+	written, err := c.Write([]byte(s))
+	if err != nil {
+		return fmt.Errorf("write auth login: %w", err)
+	}
+	if written != len(s) {
+		return fmt.Errorf("partial write %d bytes of %d", written, len(s))
+	}
+
+	return nil
+}
+
+// SendSMTP sends an email message
+func SendSMTP(msg Message) error {
+	client.Lock()
+	defer client.Unlock()
+
+	if err := client.connect(); err != nil {
+		return err
+	}
+
+	if err := client.writeln(fmt.Sprintf("MAIL FROM: %s", msg.FromAddress)); err != nil {
+		return fmt.Errorf("write mail from: %w", err)
+	}
+	if _, err := client.readln("250 Sender address accepted"); err != nil {
+		return err
+	}
+
+	if err := client.writeln(fmt.Sprintf("RCPT TO: %s", msg.ToAddress)); err != nil {
+		return fmt.Errorf("write rcpt to: %w", err)
+	}
+	if _, err := client.readln("250 Recipient address accepted"); err != nil {
+		return err
+	}
+
+	if err := client.writeln("DATA"); err != nil {
+		return err
+	}
+	if _, err := client.readln("354 "); err != nil {
+		return err
+	}
+
+	if len(msg.ToName) > 0 {
+		if err := client.writeln(fmt.Sprintf("To: %q <%s>", msg.ToName, msg.ToAddress)); err != nil {
+			return fmt.Errorf("write to: %w", err)
+		}
+	}
+	if len(msg.FromName) > 0 {
+		if err := client.writeln(fmt.Sprintf("From: %q <%s>", msg.FromName, msg.FromAddress)); err != nil {
+			return fmt.Errorf("write from: %w", err)
+		}
+	}
+	if len(msg.Subject) > 0 {
+		if err := client.writeln(fmt.Sprintf("Subject: %s", msg.Subject)); err != nil {
+			return fmt.Errorf("write subject: %w", err)
+		}
+	}
+
+	// mime multipart
+	if err := client.writeln("MIME-Version: 1.0\r\nContent-Type: multipart/alternative; boundary=c3VwYWhpbmZyYQ\r\n"); err != nil {
```

Comment:
What does the `boundary` do?

Reply:
part of how MIME formatting works, it tells the smtp server where each section starts and stops